Quartz Vertx and Aws SDK Threads not properly interrupted for DevMode live Reload

### Describe the bug

Quarkus: 3.27
Maven: 3.9.8
Java:  21.0.6 2025-01-21 LTS

I am using quartz to create a `@Scheduled` job that is using `io.quarkiverse.amazonservices: quarkus-amazon-sqs` SqsClient.receiveMessage. The local environment is using localstack with sqs.

On starting the application (`mvn compile quarkus:dev`) the job will run normally. If live-reload happens, the job restarts but the aws client aborts the receiveMessage call because of an internal check looking at `Thread.interrupted()`

The thread running the job was a vertx-worker thread. Changing the configuration 
`quarkus.quartz.run-blocking-scheduled-method-on-quartz-thread: true`
solved the issue by using the quartz threads for the job. But vertx should work as well.

### Expected behavior

Live reload should properly handle vertx threads used by quartz jobs.

### Actual behavior

Aws SDK client aborts the call to sqs because of an interrupted thread after a live-reload.

### How to Reproduce?
Using this project.

[code-with-quarkus.zip](https://github.com/user-attachments/files/24511559/code-with-quarkus.zip)

1. run ./mvnw compile quarkus:dev
2. Verify job start that the thread running the job is a vertx-worker
3. Trigger a live reload



### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk 21.0.6 2025-01-21 LTS

### Quarkus version or git rev

3.27.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.8

### Additional information

Logs from zip project above

```
__  ____  __  _____   ___  __ ____  ______
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2026-01-08 17:08:35,185 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.30.6) started in 5.771s. Listening on: http://localhost:8080
2026-01-08 17:08:35,185 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2026-01-08 17:08:35,185 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [amazon-sdk-sqs, cdi, compose, quartz, rest, scheduler, smallrye-context-propagation, vertx]
2026-01-08 17:08:35,349 INFO  [org.acme.ExampleJob] (vert.x-worker-thread-1) Example Job Complete.
2026-01-08 17:08:43,612 INFO  [io.quarkus.deployment.dev.RuntimeUpdatesProcessor] (Aesh InputStream Reader) Restarting as requested by the user.
2026-01-08 17:08:44,119 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus stopped in 0.505s
2026-01-08 17:08:44,555 INFO  [io.quarkus.devservices.common.ContainerShutdownCloseable] (build-23) Dev Services for localstack (sqs) shut down.
__  ____  __  _____   ___  __ ____  ______
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
2026-01-08 17:08:45,887 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.30.6) started in 1.765s. Listening on: http://localhost:8080
2026-01-08 17:08:45,887 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2026-01-08 17:08:45,887 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [amazon-sdk-sqs, cdi, compose, quartz, rest, scheduler, smallrye-context-propagation, vertx]
2026-01-08 17:08:45,887 INFO  [io.quarkus.deployment.dev.RuntimeUpdatesProcessor] (Aesh InputStream Reader) Live reload total time: 2.279s
2026-01-08 17:08:45,896 ERROR [io.quarkus.scheduler.common.runtime.StatusEmitterInvoker] (vert.x-worker-thread-1) Error occurred while executing task for trigger io.quarkus.quartz.runtime.QuartzSchedulerImpl$QuartzTrigger@230abee7: java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.AbortedException: Thread was interrupted
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:874)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
        at java.base/java.util.concurrent.CompletableFuture$MinimalStage.whenComplete(CompletableFuture.java:2948)
        at io.quarkus.scheduler.common.runtime.DefaultInvoker.invoke(DefaultInvoker.java:25)
        at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
        at io.quarkus.scheduler.common.runtime.StatusEmitterInvoker.invoke(StatusEmitterInvoker.java:35)
        at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
        at io.quarkus.scheduler.common.runtime.SkipConcurrentExecutionInvoker.invoke(SkipConcurrentExecutionInvoker.java:37)
        at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
        at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeComplete(DelegateInvoker.java:36)
        at io.quarkus.scheduler.common.runtime.OffloadingInvoker$2.call(OffloadingInvoker.java:54)
        at io.quarkus.scheduler.common.runtime.OffloadingInvoker$2.call(OffloadingInvoker.java:51)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$4(ContextImpl.java:192)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
        at io.vertx.core.impl.ContextImpl$1.execute(ContextImpl.java:221)
        at io.vertx.core.impl.WorkerTask.run(WorkerTask.java:56)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: software.amazon.awssdk.core.exception.AbortedException: Thread was interrupted
        at software.amazon.awssdk.core.exception.AbortedException$BuilderImpl.build(AbortedException.java:102)
        at software.amazon.awssdk.core.exception.AbortedException.create(AbortedException.java:38)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.handleInterruptedException(ApiCallTimeoutTrackingStage.java:148)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.translatePipelineException(ApiCallTimeoutTrackingStage.java:109)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:64)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:43)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:50)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:32)
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
        at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
        at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:210)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:103)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:173)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:80)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:182)
        at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:74)
        at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
        at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:53)
        at software.amazon.awssdk.services.sqs.DefaultSqsClient.receiveMessage(DefaultSqsClient.java:3340)
        at software.amazon.awssdk.services.sqs.SqsClient_dbZ6_G5adIVLg2550zQpe-OpS8M_Synthetic_ClientProxy.receiveMessage(Unknown Source)
        at org.acme.SqsHandler.receiveMessage(SqsHandler.java:25)
        at org.acme.SqsHandler_ClientProxy.receiveMessage(Unknown Source)
        at org.acme.ExampleJob.executeJob(ExampleJob.java:26)
        at org.acme.ExampleJob_ScheduledInvoker_executeJob_952519a823738725fb11b965dd917e07b656fb4d.invokeBean(Unknown Source)
        ... 22 more
Caused by: software.amazon.awssdk.core.exception.SdkInterruptedException
        at software.amazon.awssdk.core.internal.http.InterruptMonitor.checkInterrupted(InterruptMonitor.java:40)
        at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:52)
        at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:35)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:82)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:62)
        ... 43 more

```

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
@@ -109,6 +109,7 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
     private Map<DotName, Set<DotName>> classToRecompilationTargets = new HashMap<>();
 
     private final List<Runnable> preScanSteps = new CopyOnWriteArrayList<>();
+    private final List<Runnable> preRestartSteps = new CopyOnWriteArrayList<>();
     private final List<Runnable> postRestartSteps = new CopyOnWriteArrayList<>();
     private final List<Consumer<Set<String>>> noRestartChangesConsumers = new CopyOnWriteArrayList<>();
     private final List<HotReplacementSetup> hotReplacementSetup = new ArrayList<>();
@@ -554,7 +555,13 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
                 } else if (forceRestart && userInitiated) {
                     log.info("Restarting as requested by the user.");
                 }
-
+                for (Runnable step : preRestartSteps) {
+                    try {
+                        step.run();
+                    } catch (Throwable t) {
+                        log.error("Pre Restart step failed", t);
+                    }
+                }
                 restartCallback.accept(filesChanged, changedClassResults);
                 long timeNanoSeconds = System.nanoTime() - startNanoseconds;
                 log.infof("Live reload total time: %ss ", Timing.convertToBigDecimalSeconds(timeNanoSeconds));
@@ -642,6 +649,11 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
     @Override
     public void addPreScanStep(Runnable runnable) {
         preScanSteps.add(runnable);
+    }
+
+    @Override
+    public void addPreRestartStep(Runnable runnable) {
+        preRestartSteps.add(runnable);
     }
 
     @Override

--- a/core/devmode-spi/src/main/java/io/quarkus/dev/spi/HotReplacementContext.java
+++ b/core/devmode-spi/src/main/java/io/quarkus/dev/spi/HotReplacementContext.java
@@ -70,6 +70,11 @@ public interface HotReplacementContext {
     Set<String> syncState(Map<String, String> fileHashes);
 
     /**
+     * Adds a task that is run before the restart is performed.
+     */
+    void addPreRestartStep(Runnable runnable);
+
+    /**
      * Adds a task that is run after the restart is performed.
      */
     void addPostRestartStep(Runnable runnable);

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/VertxHttpHotReplacementSetup.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/VertxHttpHotReplacementSetup.java
@@ -53,7 +53,7 @@ public class VertxHttpHotReplacementSetup implements HotReplacementSetup {
                 RemoteSyncHandler.doPreScan();
             }
         });
-        hotReplacementContext.addPostRestartStep(new Runnable() {
+        hotReplacementContext.addPreRestartStep(new Runnable() {
             @Override
             public void run() {
                 // If not on a worker thread then attempt to re-initialize the dev mode executor

